src/visualisation/geoviewer.py

```diff
@@ -0,0 +1,414 @@
+"""This module contains the GeoGraphViewer to visualise GeoGraphs"""
+
+from __future__ import annotations
+from typing import TYPE_CHECKING, List, Union
+
+import folium
+import pandas as pd
+import ipywidgets as widgets
+import ipyleaflet
+import traitlets
+import logging
+
+from src.models import metrics, geograph
+from src.visualisation import (
+    folium_utils,
+    graph_utils,
+    widget_utils,
+    control_widgets,
+    style,
+)
+from src.constants import CHERNOBYL_COORDS_WGS84, WGS84
+
+if TYPE_CHECKING:
+    import geopandas as gpd
+
+
+class GeoGraphViewer(ipyleaflet.Map):
+    """Class for interactively viewing a GeoGraph."""
+
+    def __init__(
+        self,
+        center: List[int, int] = CHERNOBYL_COORDS_WGS84,
+        zoom: int = 7,
+        layout: Union[widgets.Layout, None] = None,
+        metric_list: List[str] = metrics.STANDARD_METRICS,
+        logging_level=logging.DEBUG,
+        **kwargs
+    ) -> None:
+        """Class for interactively viewing a GeoGraph.
+
+        Args:
+            center (List[int, int], optional): center of the map. Defaults to
+                CHERNOBYL_COORDS_WGS84.
+            zoom (int, optional): initial zoom level. Defaults to 7.
+            layout (Union[widgets.Layout, None], optional): layout passed to
+                ipyleaflet.Map. Defaults to None.
+            metric_list (List[str], optional): list of GeoGraph metrics to be shown.
+                Defaults to metrics.STANDARD_METRICS.
+            logging_level ([type], optional): python logging level. Defaults to
+                logging.DEBUG.
+        """
+        super().__init__(
+            center=center,
+            zoom=zoom,
+            scroll_wheel_zoom=True,
+            crs=ipyleaflet.projections.EPSG3857,  # EPSG code for WGS84 CRS
+            zoom_snap=0.1,
+            **kwargs
+        )
+        # There seems to be no easy way to add UTM35N to ipyleaflet.Map(), hence WGS84.
+        self.gpd_crs_code = WGS84
```

Comment:
What does this affect? Since WGS84 is a degrees-based CRS and UTM35N is not.

Reply:
The CRS of the viewer application always is WGS84, because ipyleaflet doesn't support UTM 35N or similar natively as far as I could spot it. So any geometry is reprojected from geo_graph.crs to WGS84 before being added as a layer (graph, components, pgons, etc.). This shouldn't affect any geograph metrics as this all happens internally in the viewer. It's not optimal, but I don't think it's adds major complexity as this is only done once on loading (add_graph()).